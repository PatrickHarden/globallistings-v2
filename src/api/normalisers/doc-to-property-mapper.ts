import { path, map, compose, prop, find, propEq, composeWith, isNil, chain } from 'ramda';
import { Maybe } from '../../types/common/maybe';
import { Hero, Charge, SizeRange, Property } from '../../types/property';

const key = prop('Common.PrimaryKey');
const safeCompose = composeWith((f: any, res: any) => (isNil(res) ? res : f(res)));

const safeComposeAndLog = (msg = 'unknown') =>
  composeWith((f: any, res: any) => {
    if (isNil(res)) {
      // TODO better error handling
      console.error(msg);
      return res;
    }
    return f(res);
  });

const composeAndLog = (propertyName: string, doc: Document) =>
  safeComposeAndLog(`Failed to map "${propertyName}" for property with id: ${key(doc)}`);

const getItemForLocale = (locale: string) => find(h => h['Common.CultureCode'] === locale);

const mapAddress = (address: CommonAddress | undefined) => {
  if (!address) {
    throw new Error('Address must be present on the property');
  }
  return {
    Country: address['Common.Country'],
    Line1: address['Common.Line1'],
    Line2: address['Common.Line2'],
    Locality: address['Common.Locallity'],
    PostCode: address['Common.PostCode'],
  };
};

const docToAddress = (doc: Document) => mapAddress(doc['Common.ActualAddress']);

const docToNumberOfBedrooms = (doc: Document) => doc['Common.NumberOfBedrooms'];

const mapLocation = (location: number[] | undefined) => {
  if (!location) {
    throw new Error('Property must have a location defined');
  }
  const [lat, lon] = location;
  return { lat, lon };
};

const docToLocation = (doc: Document) =>
  compose(
    mapLocation,
    path(['Common.GeoLocation', 'geometry', 'coordinates'])
  )(doc);

const mapContact = (contact: CommonContact) => ({
  phone: contact['Common.EmailAddress'],
  email: contact['Common.TelephoneNumber'],
  name: contact['Common.AgentName'],
});

const docToContacts = (doc: Document) =>
  composeAndLog('Common.ContactGroup', doc)([
    map(mapContact),
    path(['Common.ContactGroup', 'Common.Contacts']),
  ])(doc);

const docToHighlights = (locale: string) => (doc: Document) =>
  composeAndLog('Common.Highlights', doc)([
    map(prop('Common.Text')),
    compose(
      map(getItemForLocale(locale)),
      map(prop('Common.Highlight'))
    ),
    prop('Common.Highlights'),
  ])(doc);

const docToStrapline = (locale: string) =>
  safeCompose([prop('Common.Text'), getItemForLocale(locale), prop('Common.Strapline')]);

const docToLongDescription = (locale: string) =>
  safeCompose([prop('Common.Text'), getItemForLocale(locale), prop('Common.LongDescription')]);

const mapEnergyPerformance = safeCompose([
  map(prop('Common.Uri')),
  prop('Common.EnergyPerformanceInformation'),
]);

const mapFloorPlans = safeCompose([
  map(prop('Common.Resource.Uri')),
  chain((p: DynamicPrimaryImage) => p['Common.ImageResources']),
  prop('Common.FloorPlans'),
]);

type ImageSize = 'small' | 'large';

const getImgUri = (size: ImageSize) =>
  compose(
    (img: CommonImageResource) => img['Common.Resource.Uri'],
    find(propEq('Common.Breakpoint', size))
  );

const mapToImg = safeCompose([
  (imgs: CommonImageResource[]) => ({
    small: getImgUri('small')(imgs),
    large: getImgUri('large')(imgs),
  }),
  prop('Common.ImageResources'),
]);

const mapPhotos = safeCompose([map(mapToImg), prop('Common.Photos')]);

const getHeroPhoto = compose<Document, DynamicPrimaryImage, Hero>(
  mapToImg,
  prop('Dynamic.PrimaryImage')
);

const mapCommonChargeToCharge = (commonCharge: CommonCharge): Charge => ({
  currencyCode: commonCharge['Common.CurrencyCode'],
  year: commonCharge['Common.Year'],
  perUnit: commonCharge['Common.PerUnit'],
  amountKind: commonCharge['Common.AmountKind'],
  taxModifier: commonCharge['Common.TaxModifer'],
  interval: commonCharge['Common.Interval'],
  amount: commonCharge['Common.Amount'],
  chargeKind: commonCharge['Common.ChargeKind'],
  onApplication: commonCharge['Common.OnApplication'],
  chargeModifier: commonCharge['Common.ChargeModifer'],
});

const docToCharges = safeCompose([map(mapCommonChargeToCharge), prop('Common.Charges')]);

// TODO add tests for missing min size and max size
const docToSizeRange = (doc: Document): Maybe<SizeRange> => {
  if (!doc['Common.Sizes']) {
    return undefined;
  }

  const minSize = doc['Common.Sizes'].find(s => s['Common.SizeKind'] === 'MinimumSize');
  const maxSize = doc['Common.Sizes'].find(s => s['Common.SizeKind'] === 'TotalSize');
  if (!minSize || !maxSize) {
    return undefined;
  }

  const minSizeValue = minSize['Common.Dimensions'][0];
  const maxSizeValue = maxSize['Common.Dimensions'][0];

  return {
    min: {
      amount: minSizeValue['Common.Amount'],
      units: minSizeValue['Common.DimensionsUnits'],
    },
    max: {
      amount: maxSizeValue['Common.Amount'],
      units: maxSizeValue['Common.DimensionsUnits'],
    },
  };
};

const mapId = (doc: Document) => doc['Common.PrimaryKey'];

export const docToProperty = (locale: string) => (doc: Document): Maybe<Property> => {
  try {
    return {
      address: docToAddress(doc),
      numberOfBedrooms: docToNumberOfBedrooms(doc),
      location: docToLocation(doc),
      contacts: docToContacts(doc),
      highlights: docToHighlights(locale)(doc),
      description: {
        strapline: docToStrapline(locale)(doc),
        longDescription: docToLongDescription(locale)(doc),
      },
      info: {
        energyPerformance: mapEnergyPerformance(doc),
        floorPlans: mapFloorPlans(doc),
      },
      photos: {
        hero: getHeroPhoto(doc),
        main: mapPhotos(doc),
      },
      charges: docToCharges(doc),
      id: mapId(doc),
      size: docToSizeRange(doc),
    };
  } catch (ex) {
    console.error(`Failed mapping property: ${mapId(doc)}`);
    console.error(ex);
    return undefined;
  }
};
