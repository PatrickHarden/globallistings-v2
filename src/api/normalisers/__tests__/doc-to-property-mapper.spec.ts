import { docToProperty } from '../doc-to-property-mapper';
import property from '../../../tests/fixtures/commercial-property.json';

describe(docToProperty, () => {
  it('should map full commercial property', () => {
    expect(docToProperty('en-GB')(property as Document)).toMatchSnapshot();
  });

  [
    { path: 'Common.ActualAddress', defined: false },
    { path: 'Common.Charges', defined: true },
    { path: 'Common.ContactGroup', defined: true },
    { path: 'Common.Highlights', defined: true },
    { path: 'Common.GeoLocation', defined: false },
    { path: 'Dynamic.PrimaryImage', defined: true },
    { path: 'Common.Sizes', defined: true },
  ].forEach(td => {
    it(`should map property without ${td.path}`, () => {
      const data = { ...property, [td.path]: undefined };
      if (td.defined) {
        expect(docToProperty('en-GB')(data as any)).toMatchSnapshot();
      } else {
        expect(docToProperty('en-GB')(data as any)).toEqual(undefined);
      }
    });
  });
});
