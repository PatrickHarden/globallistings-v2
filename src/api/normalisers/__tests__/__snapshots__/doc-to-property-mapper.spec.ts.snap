// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`docToProperty should map full commercial property 1`] = `
Object {
  "address": Object {
    "Country": "UK",
    "Line1": "India House",
    "Line2": "45 Curlew Street",
    "Locality": "London",
    "PostCode": "SE1 2ND",
  },
  "charges": Array [
    Object {
      "amount": 15,
      "amountKind": "Value",
      "chargeKind": "BusinessRates",
      "chargeModifier": undefined,
      "currencyCode": "GBP",
      "interval": "Annually",
      "onApplication": undefined,
      "perUnit": "sqft",
      "taxModifier": "None",
      "year": 2019,
    },
    Object {
      "amount": 9.22,
      "amountKind": "Value",
      "chargeKind": "ServiceCharge",
      "chargeModifier": undefined,
      "currencyCode": "GBP",
      "interval": "Annually",
      "onApplication": undefined,
      "perUnit": "sqft",
      "taxModifier": "None",
      "year": 2018,
    },
    Object {
      "amount": undefined,
      "amountKind": undefined,
      "chargeKind": "Rent",
      "chargeModifier": "OnApplication",
      "currencyCode": undefined,
      "interval": undefined,
      "onApplication": true,
      "perUnit": undefined,
      "taxModifier": undefined,
      "year": undefined,
    },
  ],
  "contacts": Array [
    Object {
      "email": "+44 207 182 2577",
      "name": "Ben Deacon",
      "phone": "Ben.Deacon@cbre.com",
    },
    Object {
      "email": "+44 207 182 2165",
      "name": "Alex Kerr",
      "phone": "alex.kerr@cbre.com",
    },
  ],
  "description": Object {
    "longDescription": undefined,
    "strapline": undefined,
  },
  "highlights": Array [
    "Perimeter Comfort Cooling",
    "Raised Floors",
    "LED Lighting",
    "Commissionaire",
  ],
  "id": "GB-Plus-467051",
  "info": Object {
    "energyPerformance": undefined,
    "floorPlans": undefined,
  },
  "location": Object {
    "lat": 51.50250459999999,
    "lon": -0.07365300000003572,
  },
  "numberOfBedrooms": 0,
  "photos": Object {
    "hero": Object {
      "large": "/resources/fileassets/GB-Plus-467051/11221671/India%20House%2045%20Curlew%20Street%20Butlers%20Wharf%20-%20353892%20Ext%20land_RGB_1600px_Photo_1_large.jpg",
      "small": "/resources/fileassets/GB-Plus-467051/11221671/India%20House%2045%20Curlew%20Street%20Butlers%20Wharf%20-%20353892%20Ext%20land_RGB_1600px_Photo_1_small.jpg",
    },
    "main": undefined,
  },
  "size": Object {
    "max": Object {
      "amount": 6655,
      "units": "sqft",
    },
    "min": Object {
      "amount": 6655,
      "units": "sqft",
    },
  },
}
`;

exports[`docToProperty should map property without Common.Charges 1`] = `
Object {
  "address": Object {
    "Country": "UK",
    "Line1": "India House",
    "Line2": "45 Curlew Street",
    "Locality": "London",
    "PostCode": "SE1 2ND",
  },
  "charges": undefined,
  "contacts": Array [
    Object {
      "email": "+44 207 182 2577",
      "name": "Ben Deacon",
      "phone": "Ben.Deacon@cbre.com",
    },
    Object {
      "email": "+44 207 182 2165",
      "name": "Alex Kerr",
      "phone": "alex.kerr@cbre.com",
    },
  ],
  "description": Object {
    "longDescription": undefined,
    "strapline": undefined,
  },
  "highlights": Array [
    "Perimeter Comfort Cooling",
    "Raised Floors",
    "LED Lighting",
    "Commissionaire",
  ],
  "id": "GB-Plus-467051",
  "info": Object {
    "energyPerformance": undefined,
    "floorPlans": undefined,
  },
  "location": Object {
    "lat": 51.50250459999999,
    "lon": -0.07365300000003572,
  },
  "numberOfBedrooms": 0,
  "photos": Object {
    "hero": Object {
      "large": "/resources/fileassets/GB-Plus-467051/11221671/India%20House%2045%20Curlew%20Street%20Butlers%20Wharf%20-%20353892%20Ext%20land_RGB_1600px_Photo_1_large.jpg",
      "small": "/resources/fileassets/GB-Plus-467051/11221671/India%20House%2045%20Curlew%20Street%20Butlers%20Wharf%20-%20353892%20Ext%20land_RGB_1600px_Photo_1_small.jpg",
    },
    "main": undefined,
  },
  "size": Object {
    "max": Object {
      "amount": 6655,
      "units": "sqft",
    },
    "min": Object {
      "amount": 6655,
      "units": "sqft",
    },
  },
}
`;

exports[`docToProperty should map property without Common.ContactGroup 1`] = `
Object {
  "address": Object {
    "Country": "UK",
    "Line1": "India House",
    "Line2": "45 Curlew Street",
    "Locality": "London",
    "PostCode": "SE1 2ND",
  },
  "charges": Array [
    Object {
      "amount": 15,
      "amountKind": "Value",
      "chargeKind": "BusinessRates",
      "chargeModifier": undefined,
      "currencyCode": "GBP",
      "interval": "Annually",
      "onApplication": undefined,
      "perUnit": "sqft",
      "taxModifier": "None",
      "year": 2019,
    },
    Object {
      "amount": 9.22,
      "amountKind": "Value",
      "chargeKind": "ServiceCharge",
      "chargeModifier": undefined,
      "currencyCode": "GBP",
      "interval": "Annually",
      "onApplication": undefined,
      "perUnit": "sqft",
      "taxModifier": "None",
      "year": 2018,
    },
    Object {
      "amount": undefined,
      "amountKind": undefined,
      "chargeKind": "Rent",
      "chargeModifier": "OnApplication",
      "currencyCode": undefined,
      "interval": undefined,
      "onApplication": true,
      "perUnit": undefined,
      "taxModifier": undefined,
      "year": undefined,
    },
  ],
  "contacts": undefined,
  "description": Object {
    "longDescription": undefined,
    "strapline": undefined,
  },
  "highlights": Array [
    "Perimeter Comfort Cooling",
    "Raised Floors",
    "LED Lighting",
    "Commissionaire",
  ],
  "id": "GB-Plus-467051",
  "info": Object {
    "energyPerformance": undefined,
    "floorPlans": undefined,
  },
  "location": Object {
    "lat": 51.50250459999999,
    "lon": -0.07365300000003572,
  },
  "numberOfBedrooms": 0,
  "photos": Object {
    "hero": Object {
      "large": "/resources/fileassets/GB-Plus-467051/11221671/India%20House%2045%20Curlew%20Street%20Butlers%20Wharf%20-%20353892%20Ext%20land_RGB_1600px_Photo_1_large.jpg",
      "small": "/resources/fileassets/GB-Plus-467051/11221671/India%20House%2045%20Curlew%20Street%20Butlers%20Wharf%20-%20353892%20Ext%20land_RGB_1600px_Photo_1_small.jpg",
    },
    "main": undefined,
  },
  "size": Object {
    "max": Object {
      "amount": 6655,
      "units": "sqft",
    },
    "min": Object {
      "amount": 6655,
      "units": "sqft",
    },
  },
}
`;

exports[`docToProperty should map property without Common.Highlights 1`] = `
Object {
  "address": Object {
    "Country": "UK",
    "Line1": "India House",
    "Line2": "45 Curlew Street",
    "Locality": "London",
    "PostCode": "SE1 2ND",
  },
  "charges": Array [
    Object {
      "amount": 15,
      "amountKind": "Value",
      "chargeKind": "BusinessRates",
      "chargeModifier": undefined,
      "currencyCode": "GBP",
      "interval": "Annually",
      "onApplication": undefined,
      "perUnit": "sqft",
      "taxModifier": "None",
      "year": 2019,
    },
    Object {
      "amount": 9.22,
      "amountKind": "Value",
      "chargeKind": "ServiceCharge",
      "chargeModifier": undefined,
      "currencyCode": "GBP",
      "interval": "Annually",
      "onApplication": undefined,
      "perUnit": "sqft",
      "taxModifier": "None",
      "year": 2018,
    },
    Object {
      "amount": undefined,
      "amountKind": undefined,
      "chargeKind": "Rent",
      "chargeModifier": "OnApplication",
      "currencyCode": undefined,
      "interval": undefined,
      "onApplication": true,
      "perUnit": undefined,
      "taxModifier": undefined,
      "year": undefined,
    },
  ],
  "contacts": Array [
    Object {
      "email": "+44 207 182 2577",
      "name": "Ben Deacon",
      "phone": "Ben.Deacon@cbre.com",
    },
    Object {
      "email": "+44 207 182 2165",
      "name": "Alex Kerr",
      "phone": "alex.kerr@cbre.com",
    },
  ],
  "description": Object {
    "longDescription": undefined,
    "strapline": undefined,
  },
  "highlights": undefined,
  "id": "GB-Plus-467051",
  "info": Object {
    "energyPerformance": undefined,
    "floorPlans": undefined,
  },
  "location": Object {
    "lat": 51.50250459999999,
    "lon": -0.07365300000003572,
  },
  "numberOfBedrooms": 0,
  "photos": Object {
    "hero": Object {
      "large": "/resources/fileassets/GB-Plus-467051/11221671/India%20House%2045%20Curlew%20Street%20Butlers%20Wharf%20-%20353892%20Ext%20land_RGB_1600px_Photo_1_large.jpg",
      "small": "/resources/fileassets/GB-Plus-467051/11221671/India%20House%2045%20Curlew%20Street%20Butlers%20Wharf%20-%20353892%20Ext%20land_RGB_1600px_Photo_1_small.jpg",
    },
    "main": undefined,
  },
  "size": Object {
    "max": Object {
      "amount": 6655,
      "units": "sqft",
    },
    "min": Object {
      "amount": 6655,
      "units": "sqft",
    },
  },
}
`;

exports[`docToProperty should map property without Common.Sizes 1`] = `
Object {
  "address": Object {
    "Country": "UK",
    "Line1": "India House",
    "Line2": "45 Curlew Street",
    "Locality": "London",
    "PostCode": "SE1 2ND",
  },
  "charges": Array [
    Object {
      "amount": 15,
      "amountKind": "Value",
      "chargeKind": "BusinessRates",
      "chargeModifier": undefined,
      "currencyCode": "GBP",
      "interval": "Annually",
      "onApplication": undefined,
      "perUnit": "sqft",
      "taxModifier": "None",
      "year": 2019,
    },
    Object {
      "amount": 9.22,
      "amountKind": "Value",
      "chargeKind": "ServiceCharge",
      "chargeModifier": undefined,
      "currencyCode": "GBP",
      "interval": "Annually",
      "onApplication": undefined,
      "perUnit": "sqft",
      "taxModifier": "None",
      "year": 2018,
    },
    Object {
      "amount": undefined,
      "amountKind": undefined,
      "chargeKind": "Rent",
      "chargeModifier": "OnApplication",
      "currencyCode": undefined,
      "interval": undefined,
      "onApplication": true,
      "perUnit": undefined,
      "taxModifier": undefined,
      "year": undefined,
    },
  ],
  "contacts": Array [
    Object {
      "email": "+44 207 182 2577",
      "name": "Ben Deacon",
      "phone": "Ben.Deacon@cbre.com",
    },
    Object {
      "email": "+44 207 182 2165",
      "name": "Alex Kerr",
      "phone": "alex.kerr@cbre.com",
    },
  ],
  "description": Object {
    "longDescription": undefined,
    "strapline": undefined,
  },
  "highlights": Array [
    "Perimeter Comfort Cooling",
    "Raised Floors",
    "LED Lighting",
    "Commissionaire",
  ],
  "id": "GB-Plus-467051",
  "info": Object {
    "energyPerformance": undefined,
    "floorPlans": undefined,
  },
  "location": Object {
    "lat": 51.50250459999999,
    "lon": -0.07365300000003572,
  },
  "numberOfBedrooms": 0,
  "photos": Object {
    "hero": Object {
      "large": "/resources/fileassets/GB-Plus-467051/11221671/India%20House%2045%20Curlew%20Street%20Butlers%20Wharf%20-%20353892%20Ext%20land_RGB_1600px_Photo_1_large.jpg",
      "small": "/resources/fileassets/GB-Plus-467051/11221671/India%20House%2045%20Curlew%20Street%20Butlers%20Wharf%20-%20353892%20Ext%20land_RGB_1600px_Photo_1_small.jpg",
    },
    "main": undefined,
  },
  "size": undefined,
}
`;

exports[`docToProperty should map property without Dynamic.PrimaryImage 1`] = `
Object {
  "address": Object {
    "Country": "UK",
    "Line1": "India House",
    "Line2": "45 Curlew Street",
    "Locality": "London",
    "PostCode": "SE1 2ND",
  },
  "charges": Array [
    Object {
      "amount": 15,
      "amountKind": "Value",
      "chargeKind": "BusinessRates",
      "chargeModifier": undefined,
      "currencyCode": "GBP",
      "interval": "Annually",
      "onApplication": undefined,
      "perUnit": "sqft",
      "taxModifier": "None",
      "year": 2019,
    },
    Object {
      "amount": 9.22,
      "amountKind": "Value",
      "chargeKind": "ServiceCharge",
      "chargeModifier": undefined,
      "currencyCode": "GBP",
      "interval": "Annually",
      "onApplication": undefined,
      "perUnit": "sqft",
      "taxModifier": "None",
      "year": 2018,
    },
    Object {
      "amount": undefined,
      "amountKind": undefined,
      "chargeKind": "Rent",
      "chargeModifier": "OnApplication",
      "currencyCode": undefined,
      "interval": undefined,
      "onApplication": true,
      "perUnit": undefined,
      "taxModifier": undefined,
      "year": undefined,
    },
  ],
  "contacts": Array [
    Object {
      "email": "+44 207 182 2577",
      "name": "Ben Deacon",
      "phone": "Ben.Deacon@cbre.com",
    },
    Object {
      "email": "+44 207 182 2165",
      "name": "Alex Kerr",
      "phone": "alex.kerr@cbre.com",
    },
  ],
  "description": Object {
    "longDescription": undefined,
    "strapline": undefined,
  },
  "highlights": Array [
    "Perimeter Comfort Cooling",
    "Raised Floors",
    "LED Lighting",
    "Commissionaire",
  ],
  "id": "GB-Plus-467051",
  "info": Object {
    "energyPerformance": undefined,
    "floorPlans": undefined,
  },
  "location": Object {
    "lat": 51.50250459999999,
    "lon": -0.07365300000003572,
  },
  "numberOfBedrooms": 0,
  "photos": Object {
    "hero": undefined,
    "main": undefined,
  },
  "size": Object {
    "max": Object {
      "amount": 6655,
      "units": "sqft",
    },
    "min": Object {
      "amount": 6655,
      "units": "sqft",
    },
  },
}
`;
